contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+
+# Design Requirements
+
+If we consider the use cases listed above, the main design requirements can roughly be listed as:
+
+1 – An unique, consistent and discoverable identity for each replica/instance across the
+federated clusters.
+
+2 – It must be possible for the federated statefulset to first safely form an initial quorum
+(probably using cluster local replicas and identities).
+It should further also be possible to add members (probably from other clusters and global
+identities) to this initial quorum.
+
+3 – The ability to scale, across clusters in some deterministic fashion.
+
+4 – An extended design requirement (which might not be a necessary requirement for this
+phase for work) can be an ability to migrate a pod from one node (in one cluster) to
+another node (possibly in another cluster), with stable network identity along with its attached volumes.
+
+# Design Proposals
+
+The Statefulset object and APIs are already defined in k8s.
+The same needs to be exposed from federation to ensure that the interfaces remain consistent
+with a stand-alone k8s cluster.
+A separate controller at federation would however be useful.
+
+We considered two alternative designs for this proposal, both of which are further described
+in this document. The detailed design for the primary alternative is elaborated in the next section.
+A design alternative is also evaluated and listed later in the doc.
+
+## Detailed Design (Primary Alternative)
+
+In this approach, the federated statefulset controller will partly behave quite similar to federated
+replicaset or the federated deployment controller.
+The federated controller would create and monitor individual statefulsets partitioning and distributing
+the total stateful replicas across the federated clusters.
+
+As an additional proposal in this design we suggest the possibility of the pods having multiple identities.
+This facilitates application pods being able to participate in different quora, and leaves a
+choice with application replicas to join either the local or the global quora, as need be.
+As of now we can envision the need of two possible identities for each stateful pod (relevant to use case 2),
+enabled by this approach, but the same design can be extended for a stateful replica to get more then 2
+identities also, in future, if the need be.
+The ordinal number assigned to the pod will be calculated and assigned in exactly the same way,
+it is assigned in the k8s cluster as described
+[here](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity) and will
+be unique.
+The details of how the replicas will be distributed across federated clusters is elaborated in
+&#39;Replica distribution&#39; section below.
+The details of the multiple dns identities for each pod is separately elaborated in
+&#39;instance identity and discovery&#39; sections below.
+
+For this phase of implementation, it is proposed that the statefulset need not guarantee the
```

Comment:
I think there has been some discussion is app sig, and given the complexity of the same, I strongly think that this can be relaxed (atleast for the current phase of impl). https://github.com/kubernetes/community/pull/437#pullrequestreview-25676538